Add supplies to restock vending machines.

Vordenburg opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

Adds the ability to restock vending machines via boxes that can be ordered through the cargo ordering system. There's also a couple random spawners for the food and drink boxes. These boxes can be smashed open too to get at some of their contents.

Vending machines require their access panels to be opened first (with something like a screwdriver), then it takes a few seconds for the machine to be restocked by hitting it with the restock box.

The restock box will give you a popup if either the access panel is closed or you're using the wrong restock box.

Addresses #3857, #11580

**Screenshots**
<!-- If applicable, add screenshots to showcase your PR. If your PR is a visual change, add
screenshots or it's liable to be closed by maintainers. -->

![image](https://user-images.githubusercontent.com/114301317/200393089-61854e40-68f2-4c7a-948c-5204d57a38cd.png)

**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged
There are 4 icons for changelog entries: add, remove, tweak, fix. I trust you can figure out the rest.

You can put your name after the :cl: symbol to change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB

Generally, only put things in changelogs that players actually care about. Stuff like "Refactored X system, no changes should be visible" shouldn't be on a changelog.

For writing actual entries, don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers
-->

:cl: Vordenburg & themias
- add: Added vending machine restock crates orderable through cargo.

mirrorcult commented:
this is pretty much a duplicate of #11502. not really sure what to do here

themias commented:
> this is pretty much a duplicate of #11502. not really sure what to do here

This PR looks better than mine TBH. All I'd ask for (maybe in a future PR) is to add a pack for the cigarette machine.
(All the others mine adds are for machines that rarely run out anyway and I just added since there were sprites for them.)

theashtronaut commented:
I'd like to be able to restock all vending machines as a possibility, I also kinda like the tg restock module sprites tbh from the other pr

can be an additional PR to add the missing machines after this one tho

mirrorcult commented on Content.Server/Destructible/Thresholds/Behaviors/DumpRestockInventory.cs:
```diff
@@ -0,0 +1,68 @@
+
+            var position = system.EntityManager.GetComponent<TransformComponent>(owner).MapPosition;
+
+            var getRandomVector = () => new Vector2(system.Random.NextFloat(-Offset, Offset), system.Random.NextFloat(-Offset, Offset));
```
`system.Random.NextVector2` exists iirc

mirrorcult commented on Content.Server/VendingMachineRestockPackage/VendingMachineRestockPackageComponent.cs:
```diff
@@ -0,0 +1,38 @@
+        ///     Sound that plays when starting to restock a machine.
+        /// </summary>
+        [DataField("soundRestockStart")]
+        // Credit to: https://freesound.org/people/Defaultv/sounds/534362/
```
these shoudl be attributed in the folder with the audio files themselves, under an `attributions.yml` RGA format; there are a couple examples of this you can copy

mirrorcult commented on Content.Server/VendingMachineRestockPackage/VendingMachineRestockPackageComponent.cs:
```diff
@@ -0,0 +1,38 @@
+        ///     Sound that plays when finished restocking a machine.
+        /// </summary>
+        [DataField("soundRestockDone")]
+        // Credit to: https://freesound.org/people/felipelnv/sounds/153298/
```
same here

mirrorcult commented on Content.Server/VendingMachineRestockPackage/VendingMachineRestockPackageSystem.cs:
```diff
@@ -0,0 +1,159 @@
+            if (!TryComp<VendingMachineComponent>(args.Target, out var machineComponent))
+                return;
+
+            if (!TryMatchPackageToMachine(uid, component, machineComponent, args.User, args.Target.GetValueOrDefault()))
+                return;
+
+            if (!TryAccessMachine(uid, component, machineComponent, args.User, args.Target.GetValueOrDefault()))
+                return;
```
```suggestion
            if (!TryMatchPackageToMachine(uid, component, machineComponent, args.User, args.Target.Value))
                return;

            if (!TryAccessMachine(uid, component, machineComponent, args.User, args.Target.Value))
                return;
```

target was already null checked, so this works fine

mirrorcult reviewed (changes requested):
the sprites here are a little strange and not what i would expect. /tg/ et al has sprites for these already (and i thought we did too) that look a bit more consistent. i believe themias' duplicate PR used these, so you should look at that

mirrorcult commented:
logic + test looks good to me

metalgearsloth commented:
You coming back to this @Vordenburg ?

Vordenburg commented:
@metalgearsloth If I have free time today, I'll see to wrapping this up.

Vordenburg commented:
I'm almost complete with this. @metalgearsloth 

I've integrated everything @themias wanted originally, used the sprites requested by @mirrorcult after cleaning them up, added a few more restocks and sprites to go with them. I added a new test and cleaned the originals up based on experience I've learned since originally opening this PR.

Everything appears to be working so far. I want to review the cargo values for the restock boxes, maybe add a new test or two, and give it one final pass, but this is about ready to go soon.

I'll have to edit the original PR to add a proper changelog entry too. Should have this all done within 24hr.

metalgearsloth commented on Content.Server/Destructible/Thresholds/Behaviors/DumpRestockInventory.cs:
```diff
@@ -0,0 +1,66 @@
+            if (!system.PrototypeManager.TryIndex(randomInventory, out VendingMachineInventoryPrototype? packPrototype))
+                return;
+
+            var position = system.EntityManager.GetComponent<TransformComponent>(owner).MapPosition;
```
You already have the transform above.

metalgearsloth commented on Content.Server/Destructible/Thresholds/Behaviors/DumpRestockInventory.cs:
```diff
@@ -0,0 +1,66 @@
+
+            packagecomp.CanRestock.CopyTo(ary);
+
+            var randomInventory = ary[system.Random.Next(ary.Length)];
```
Also we have an IoC for random, same with below.

metalgearsloth commented on Content.Server/VendingMachines/Restock/VendingMachineRestockComponent.cs:
```diff
@@ -0,0 +1,36 @@
+    {
+        public CancellationTokenSource? CancelToken;
+
+        /// <summary>
+        /// The time (in seconds) that it takes to restock a machine.
+        /// </summary>
+        [DataField("restockDelay")]
+        public float RestockDelay = 8.0f;
+
+        /// <summary>
+        /// What sort of machine inventory does this restock?
+        /// This is checked against the VendingMachineComponent's pack value.
+        /// </summary>
+        [DataField("canRestock")]
+        public HashSet<string> CanRestock = new();
+
+        /// <summary>
+        ///     Sound that plays when starting to restock a machine.
+        /// </summary>
+        [DataField("soundRestockStart")]
+        public SoundSpecifier SoundRestockStart = new SoundPathSpecifier("/Audio/Machines/vending_restock_start.ogg");
+
+        /// <summary>
+        ///     Sound that plays when finished restocking a machine.
+        /// </summary>
+        [DataField("soundRestockDone")]
+        public SoundSpecifier SoundRestockDone = new SoundPathSpecifier("/Audio/Machines/vending_restock_done.ogg");
```
VV readwrite.

metalgearsloth commented on Content.Server/VendingMachines/Restock/VendingMachineRestockSystem.cs:
```diff
@@ -0,0 +1,195 @@
+
+            _doAfterSystem.DoAfter(new DoAfterEventArgs(args.User, component.RestockDelay, component.CancelToken.Token, target: args.Target)
+            {
+                BroadcastFinishedEvent = new RestockSuccessfulEvent(args.User, (EntityUid) args.Target, component.Owner),
+                BroadcastCancelledEvent = new RestockCancelledEvent(component.Owner),
```
Why aren't these directed?

metalgearsloth commented on Content.Server/VendingMachines/Restock/VendingMachineRestockSystem.cs:
```diff
@@ -0,0 +1,195 @@
+                .WithVolume(-2f)
+                .WithVariation(0.2f));
+
+            QueueDel(ev.Package);
```
How come it's not just a Del?

metalgearsloth commented on Content.Server/VendingMachines/Restock/VendingMachineRestockSystem.cs:
```diff
@@ -0,0 +1,195 @@
+            QueueDel(ev.Package);
+        }
+
+        private sealed class RestockCancelledEvent : EntityEventArgs
+        {
+            public EntityUid Package;
+
+            public RestockCancelledEvent(EntityUid package)
+            {
+                Package = package;
+            }
+        }
+
+        private sealed class RestockSuccessfulEvent : EntityEventArgs
+        {
+            public EntityUid User;
+            public EntityUid Target;
+            public EntityUid Package;
+            public RestockSuccessfulEvent(EntityUid user, EntityUid target, EntityUid package)
+            {
+                User = user;
+                Target = target;
+                Package = package;
+            }
+        }
```
Should be struct events.

metalgearsloth commented on Content.Shared/VendingMachines/SharedVendingMachineComponent.cs:
```diff
@@ -97,4 +97,6 @@ public enum EjectWireKey : byte
     }
 
     public sealed class VendingMachineSelfDispenseEvent : InstantActionEvent { };
+
+    public sealed class VendingMachineRestockEvent : HandledEntityEventArgs { };
```
Comment?

I would've assumed this is raised when it's successful but then I saw it inherits from handled.

metalgearsloth commented on Content.Shared/VendingMachines/SharedVendingMachineSystem.cs:
```diff
@@ -65,27 +75,37 @@ public List<VendingMachineInventoryEntry> GetAvailableInventory(EntityUid uid, S
+
+        foreach (var (id, amount) in entries)
+        {
+            if (_prototypeManager.HasIndex<EntityPrototype>(id))
+            {
+                if (inventory.TryGetValue(id, out VendingMachineInventoryEntry? entry))
+                    // Prevent a machine's stock from going over three times
+                    // the prototype's normal amount. This is an arbitrary
+                    // number and meant to be a convenience for someone
+                    // restocking a machine who doesn't want to force vend out
+                    // all the items just to restock one empty slot without
+                    // losing the rest of the restock.
+                    entry.Amount = Math.Min(entry.Amount + amount, 3 * amount);
+                else
+                    inventory.Add(id, new VendingMachineInventoryEntry(type, id, amount));
+            }
```
I would think the restock is just bringing it up to the normal amount.

Vordenburg replied:
If the machine is not entirely empty, you lose the difference between the starting inventory and that amount. Is that not undesirable? It would be a waste of money.

Sometimes not all items are vended out when a restock is needed. It encourages players to vend out what items haven't been completely emptied, if those items are still desirable. This allows them to restock extra within a certain limit for the sake of convenience.

metalgearsloth commented on Resources/Prototypes/Catalog/Fills/Crates/vending.yml:
```diff
@@ -0,0 +1,134 @@
+      - id: VendingMachineRestockSmokes
+        amount: 2
+
+- type: entity
+  id: CrateVendingMachineRestockSnacksFilled
+  parent: CratePlastic
+  components:
+  - type: StorageFill
+    contents:
+      - id: VendingMachineRestockChang
+      - id: VendingMachineRestockDiscountDans
+      - id: VendingMachineRestockDonut
+      - id: VendingMachineRestockGetmoreChocolateCorp
```
That is a significant amount of restocks.

Vordenburg commented on Content.IntegrationTests/Tests/VendingMachineRestockTest.cs:
```diff
@@ -0,0 +1,372 @@
+
+            await server.WaitAssertion(() =>
+            {
+                foreach (var proto in prototypeManager.EnumeratePrototypes<EntityPrototype>())
+                {
+                    if (!proto.TryGetComponent<VendingMachineRestockComponent>(out var package))
+                        continue;
+
+                    foreach(var vendingInventory in package.CanRestock)
+                    {
+                        Assert.That(prototypeManager.HasIndex<VendingMachineInventoryPrototype>(vendingInventory),
+                            $"Unknown VendingMachineInventoryPrototype {vendingInventory} on VendingMachineRestockComponent for {proto.ID}");
+                    }
+                }
+            });
```
Are there any examples or documentation for what I need to know in order to do this?

mirrorcult commented:
needs addressed reviews and can merge
